fsharp-backend/tests/Tests/Serialization.Tests.fs

```diff
@@ -0,0 +1,999 @@
+
+
+module BinarySerializationRoundtripTests =
+  let toplevelRoundtripTest =
+    testMany
+      "serializeToplevels"
+      (fun tl ->
+        let tlid = PT.Toplevel.toTLID tl
+        tl
+        |> BinarySerialization.serializeToplevel
+        |> BinarySerialization.deserializeToplevel tlid
+        |> (=) tl)
+      (List.map (fun x -> x, true) Values.testToplevels)
+
+
+  let oplistRoundtripTest =
+    test "roundtrip oplists" {
+      let actual =
+        Values.testOplist
+        |> BinarySerialization.serializeOplist 0UL
+        |> BinarySerialization.deserializeOplist 0UL
+      Expect.equal actual Values.testOplist ""
+    }
+
+module GenericSerializersTests =
+  // We've annotated each type that we serialize in Dark, with the generic serializer
+  // that it allows. Now we want to verify that the format doesn't change by having a
+  // sample value of each type, and verifying the serialized sample value doesn't
+  // change.  This gives us extremely high confidence that we're not breaking
+  // anything when we change the serializer.
+  // - If there's a change to the format, a new file will be generated.
+  //   What to do here will depend on the change, how much data is saved with the old
+  //   change, etc. You may want to save both formats and check that both are
+  //   deserializable.
+  // - If there's a change to the sample value, generate a new file.
+  //   Commit it after verifying that the format is appropriate.
+
+
+  module SampleData =
+    open Values
+
+    // Stores a list of sample outputs for `$"{serializerName}-{typeName}"
+    let data = Dictionary.T<string, List<string * string>>()
+
+    let keyFor (typeName : string) (serializerName : string) : string =
+      $"{serializerName}-{typeName}"
+
+    let fileNameFor
+      (typeName : string)
+      (serializerName : string)
+      (dataName : string)
+      : string =
+      let original = $"{serializerName}-{typeName}-{dataName}"
+      let replaced = Regex.Replace(original, "[^-_a-zA-Z0-9]", "-")
+      Regex.Replace(replaced, "[-]+", "-") + ".json"
+
+    let get
+      (typeName : string)
+      (serializerName : string)
+      (reason : string)
+      : List<string * string> =
+      data
+      |> Dictionary.get (keyFor typeName serializerName)
+      |> Exception.unwrapOptionInternal
+           "testCases should exist for allowed serializable type"
+           [ "typeName", typeName :> obj
+             "serializer", serializerName
+             "reason", reason ]
+
+
+    let generate<'v>
+      (serializerName : string)
+      (f : 'v -> string)
+      (name : string)
+      (v : 'v)
+      =
+      let typeName = string typeof<'v>
+      let key = keyFor typeName serializerName
+      let currentValue = Dictionary.get key data |> Option.unwrap []
+      let newValue = (name, f v) :: currentValue
+      data[key] <- newValue
+
+    let v<'t when 't : equality> (dataName : string) (data : 't) =
```

Comment:
I think it's a marginal cost to rename these fns to `vanilla` and `ocamlcompat` for sake of legibility

Reply:
or better yet, `genVanilla` and `genOcaml`. Even `genV` and `genO` would be better